src/software/test_util/test_util.cpp

```diff
@@ -102,4 +102,50 @@ namespace Test
         return Robot(robot_id_counter++, pt, Vector(), Angle(), AngularVelocity(),
                      Timestamp());
     }
+
+    bool TestUtil::checkGeometryEqualWithTolerance(const Polygon &poly1,
+                                                   const Polygon &poly2, double tolerance)
+    {
+        auto ppoints1 = poly1.getPoints();
+        auto ppoints2 = poly2.getPoints();
+        if (ppoints1.size() != ppoints2.size())
+        {
+            return false;
+        }
+
+        for (int i = 0; i < ppoints1.size(); i++)
+        {
+            if (!checkGeometryEqualWithTolerance(ppoints1[i], ppoints2[i], tolerance))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    bool TestUtil::checkGeometryEqualWithTolerance(const Circle &circle1,
+                                                   const Circle &circle2,
+                                                   double tolerance)
+    {
+        if (!checkGeometryEqualWithTolerance(circle1.getOrigin(), circle2.getOrigin(),
+                                             tolerance))
+        {
+            return false;
+        }
+        if (fabs(circle1.getRadius() - circle2.getRadius()) > tolerance)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    bool TestUtil::checkGeometryEqualWithTolerance(const Point &point1,
+                                                   const Point &point2, double tolerance)
+    {
+        return (point1.distanceFromPoint(point2) <=
+                (sqrt(2) * tolerance + GeomConstants::FIXED_EPSILON));
```

Comment:
Where's the `sqrt(2)` coming from?

Reply:
yeah it's a bit arbitrary, I changed it to compare each difference in x/y value to the tolerance instead